src/utils/spanUtils.ts

```diff
@@ -0,0 +1,156 @@
+    data.resultCode = String(span.attributes['http.status_code']);
+  }
+
+  if (span.attributes['http.url']) {
+    const url = new URL(span.attributes['http.url'] as string);
+    data.target = url.hostname;
+    data.data = url.href;
+
+    if (span.attributes['http.method']) {
+      data.name = `${span.attributes['http.method']} ${url.pathname}`;
+    }
+  }
+
+  return data;
+}
+
+function createRequestData(span: ReadableSpan): RequestData {
+  const data = new RequestData();
+  data.name = span.name;
+  data.id = `|${span.spanContext.traceId}.${span.spanContext.spanId}.`;
+  data.success = span.status.code === CanonicalCode.OK;
+  data.responseCode = String(span.status.code);
+  data.duration = String(hrTimeToMilliseconds(span.duration));
+  data.ver = 1;
+  data.source = undefined;
+
+  if (span.attributes['http.method']) {
+    data.name = span.attributes['http.method'] as string;
+
+    if (span.attributes['http.status_code']) {
+      data.responseCode = String(span.attributes['http.status_code']);
+    }
+
+    if (span.attributes['http.url']) {
+      data.url = span.attributes['http.url'] as string;
+    }
+
+    if (span.attributes['http.route']) {
+      data.name = `${span.attributes['http.method']} ${span.attributes['http.route']}`;
+    } else if (span.attributes['http.url']) {
+      const url = new URL(span.attributes['http.url'] as string);
+      data.name = `${span.attributes['http.method']} ${url.pathname}`;
+    }
+  }
+
+  // @todo: parse grpc attributes
+
+  return data;
+}
+
+function createInProcData(span: ReadableSpan): RemoteDependencyData {
+  const data = createDependencyData(span);
+  data.type = 'InProc';
+  data.success = true;
+  return data;
+}
+
+export function readableSpanToEnvelope(
+  span: ReadableSpan,
+  instrumentationKey: string,
+  logger?: Logger,
+): Envelope {
+  const envelope = new Envelope();
+  envelope.data = new Base();
+  const tags = createTagsFromSpan(span);
+  const properties = createPropertiesFromSpan(span);
+  let data;
+  switch (span.kind) {
+    case SpanKind.CLIENT:
+    case SpanKind.PRODUCER:
+      envelope.name = 'Microsoft.ApplicationInsights.RemoteDependency';
+      envelope.data.baseType = 'RemoteDependencyData';
+      data = createDependencyData(span);
+      break;
+    case SpanKind.SERVER:
+    case SpanKind.CONSUMER:
+      envelope.name = 'Microsoft.ApplicationInsights.Request';
+      envelope.data.baseType = 'RequestData';
+      data = createRequestData(span);
+      break;
+    case SpanKind.INTERNAL:
+      envelope.data.baseType = 'RemoteDependencyData';
+      envelope.name = 'Microsoft.ApplicationInsights.RemoteDependency';
+      data = createInProcData(span);
+      break;
+    default:
+      // never
+      if (logger) {
+        logger.error(`Unsupported span kind ${span.kind}`);
+      }
+      throw new Error(`Unsupported span kind ${span.kind}`);
+  }
+
+  envelope.data.baseData = { ...data, properties };
```

Comment:
Where is AI specific context information stored? (ai.cloud.role, ai.internal.sdkversion, etc)

Reply:
Will add in separate PR